### PR TITLE
doc strings type and add trait inheritance section

### PR DESCRIPTION
**Pull request by Tokazama**
* Add consistent type return annotation to doc strings (where
  appropriate)
* Section on inheriting behavior using `parent_type`.
* Put doc strings in manual after text

### CODE REVIEW

File: docs/src/index.md [hunk 1]
```diff
@@ -70,6 +70,29 @@ Therefore, the recommended approach for supporting static sizing in newly define
...
 Static information related to subtypes of `AbstractRange` include `known_length`, `known_first`, `known_step`, and `known_last`.
 
+## Dimensions
+
+Methods such as `size(x, dim)` need to map `dim` to the dimensions of `x`.
+Typically, `dim` is an `Int` with an invariant mapping to the dimensions of `x`.
+Some methods accept `:` or a tuple of dimensions as an argument.
+`ArrayInterface` also considers `StaticInt` a viable dimension argument.
+
+[`ArrayInterface.to_dims`](@ref) helps ensure that `dim` is converted to a viable dimension mapping helps with type stability.
```
Comment (chriselrod) [inline docs/src/index.md:80]: > helps ensure that `dim` is converted to a viable dimension mapping helps with type stability.

"in a manner that helps with type stability"?

File: docs/src/index.md [hunk 2]
```diff
@@ -70,6 +70,29 @@ Therefore, the recommended approach for supporting static sizing in newly define
...
+For example, all `Integers` passed to `to_dims` are converted to an `Int` (unless `dim` is a `StaticInt`).
```
Comment (chriselrod) [inline docs/src/index.md:81]: "converted to `Int`"?

File: docs/src/index.md [hunk 3]
```diff
@@ -70,6 +70,29 @@ Therefore, the recommended approach for supporting static sizing in newly define
...
+
+Methods accepting dimension specific arguments should use some variation of the following pattern.
+
+```julia
+f(x, dim) = f(x, ArrayInterface.to_dims(x, dim))
+f(x, dim::Int) = ...
+f(x, dim::StaticInt) = ...
+```
+
+If `x` has a dimension named `:dim_1` then calling `f(x, :dim_1)` would result in `f(x, 1)`.
```
Comment (chriselrod) [inline docs/src/index.md:93]: Someone could conceivably name dims `dim_0, dim_1` or `dim_2, dim_1`, etc.
But what is meant is probably obvious enough in context?

Comment (Tokazama) [reply docs/src/index.md]: No reason to make people guess, so I fixed it.

**Review by chriselrod — APPROVED** [review]
Looks great!

**Pull request merged**